cpp/include/data_structure/tree/README.md

```diff
@@ -34,3 +35,48 @@ _Update_    | _O(log N)_   | _O(N)_
 _Calculate_ | _O(log N)_   | _O(N)_
 
 Where N is the number of elements in the original array.
+
+## 2. Binary search tree
+A binary tree where the value of each node to the right of the current node is larger than or equal to the value of the current node, and the value of the node to the left is less than or equal to the the value of the current node. Storing data using a binary search tree allows a good compromise to be made between insertion speed and search speed.
+
+### Usage
+```c++
+BinarySearchTree tree;
+
+// inserting values into the tree
+tree.insert(1);
+tree.insert(0);
+tree.insert(-1);
+tree.insert(5);
+tree.insert(2);
+tree.insert(3);
+tree.insert(4);
+tree.insert(6);
+
+// to delete an element
+tree.remove(4);
+
+// to search for an element
+tree.search(2)
+
+// for inorder traversal using the recursive method
+// Output will be this array: 
+tree.traversal_inorder_recursive();
+
+// for inorder traversal using the recursive method
+// Output will be this array: 
+tree.traversal_inorder_iterative();
+
+// the rest of the traversals follow the same method
+// replacing inorder with preorder or postorder
+```
+
+### Complexity
+Operation                | Time         | Space
+------------------------ | ------------ |-------------------
+_Insert_                 | _O(log N)_   | _O(1)_
+_Remove_                 | _O(log N)_   | _O(1)_
+_Search_                 | _O(log N)_   | _O(1)_
```

Comment:
The worst-case time complexity of these operations is _O(N)_. We should mention this too somehow. One way of doing it (the columns can be aligned but it's not required):

```
Operation | Time | Space
--------- | ---- | -----
Insert | Average: _O(log N)_<br> Worst: _O(N)_ | _O(1)_
Remove | Average: _O(log N)_<br> Worst: _O(N)_ | _O(1)_
...
```

Operation | Time | Space
-- | -- | --
Insert | Average: _O(log N)_<br> Worst: _O(N)_ | _O(1)_
Remove | Average: _O(log N)_<br> Worst: _O(N)_ | _O(1)_
...

The downside is that the plaintext readability is reduced due to the use of HTML tags.

Another way:

```
Operation | Time | Space
--------- | ---- | -----
Insert | Average: O(log N), Worst: O(N) | O(1)
Remove | Average: O(log N), Worst: O(N) | O(1)
...
```

Operation | Time | Space
--------- | ---- | -----
Insert | Average: O(log N), Worst: O(N) | O(1)
Remove | Average: O(log N), Worst: O(N) | O(1)
...

This does not look as good in the rendered output but is more readable in the plaintext form.

Reply:
Isn't the worst-case log<sub>2</sub>(N), since in the worst case the search would end with the element being at the bottom of the tree. Even if it is at the bottom of the tree we never had to compare all the elements. 

This is a Wikipedia picture of it: https://en.wikipedia.org/wiki/Binary_search_algorithm#/media/File:Binary_search_complexity.svg.

Reply:
That's the average-case complexity. The worst-case complexity is when you insert elements that are all either in increasing or decreasing order and get a skewed binary tree (with just one long branch), where searching for the last element will require traversing all the elements, leading to O(N) time complexity.

<img src="https://qphs.fs.quoracdn.net/main-qimg-eccb472654af69385afe691e9958a713.webp" width="400"/>

Reply:
I completely forgot about skewed trees. Sorry about that. I should complete this later this week.

Reply:
I've changed it. Sorry about all the trouble and thanks for the help!